internal/provider/resource_gitlab_project_protected_environment_test.go

```diff
@@ -0,0 +1,276 @@
+					}),
+				),
+			},
+			// Update the Protected Environment to get back to initial settings
+			{
+				SkipFunc: isRunningInCE,
+				Config:   testAccGitlabProjectProtectedEnvironmentConfig(rInt, wildcard),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGitlabProjectProtectedEnvironmentExists("gitlab_project_protected_environment.ProjectProtectedEnvironment", &pt),
+					testAccCheckGitlabProjectProtectedEnvironmentAttributes(&pt, &testAccGitlabProjectProtectedEnvironmentExpectedAttributes{
+						Name:              fmt.Sprintf("ProjectProtectedEnvironment-%d%s", rInt, wildcard),
+						CreateAccessLevel: accessLevelValueToName[gitlab.DeveloperPermissions],
+					}),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGitlabProjectProtectedEnvironmentExists(n string, pt *gitlab.ProtectedEnvironment) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not Found: %s", n)
+		}
+		project, environment, err := parseTwoPartID(rs.Primary.ID)
+		if err != nil {
+			return fmt.Errorf("Error in Splitting Project ID and Environment Name")
+		}
+
+		pes, _, err := testGitlabClient.ProtectedEnvironments.ListProtectedEnvironments(project, nil)
+		if err != nil {
+			return err
+		}
+		for _, gotpe := range pes {
+			if gotpe.Name == environment {
+				*pt = *gotpe
+				return nil
+			}
+		}
+		return fmt.Errorf("Protected Environment does not exist")
+	}
+}
+
+type testAccGitlabProjectProtectedEnvironmentExpectedAttributes struct {
+	Name              string
+	CreateAccessLevel string
+}
+
+func testAccCheckGitlabProjectProtectedEnvironmentAttributes(pt *gitlab.ProtectedEnvironment, want *testAccGitlabProjectProtectedEnvironmentExpectedAttributes) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		if pt.Name != want.Name {
+			return fmt.Errorf("got name %q; want %q", pt.Name, want.Name)
+		}
+
+		if pt.DeployAccessLevels[0].AccessLevel != accessLevelNameToValue[want.CreateAccessLevel] {
+			return fmt.Errorf("got Create access levels %q; want %q", pt.DeployAccessLevels[0].AccessLevel, accessLevelNameToValue[want.CreateAccessLevel])
+		}
+
+		return nil
+	}
+}
+
+func testAccCheckGitlabProjectProtectedEnvironmentDestroy(s *terraform.State) error {
+	var project string
+	var environment string
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type == "gitlab_project" {
+			project = rs.Primary.ID
+		} else if rs.Type == "gitlab_project_protected_environment" {
+			environment = rs.Primary.ID
+		}
+	}
+
+	pt, response, err := testGitlabClient.ProtectedEnvironments.GetProtectedEnvironment(project, environment)
+	if err == nil {
+		if pt != nil {
+			return fmt.Errorf("project Protected Environment %s still exists", environment)
+		}
+	}
+	if response != nil && response.StatusCode != 404 {
+		return err
+	}
+	return nil
+}
+
+func testAccGitlabProjectProtectedEnvironmentConfig(rInt int, postfix string) string {
```

Comment:
Generally in the tests, please only include what is actually under test in the terraform config, e.g. here the `gitlab_project_protected_environment` resource. Everything else should be done via helper functions, see: https://github.com/gitlabhq/terraform-provider-gitlab/blob/main/internal/provider/helper_test.go#L102

and some examples here: 

https://github.com/gitlabhq/terraform-provider-gitlab/blob/375586b509db071c8aaf9f4f8b8f6d4ca6bd6359/internal/provider/data_source_gitlab_project_issues_test.go#L13

For you interesting are:

* [`testAccCreateProject`](https://github.com/gitlabhq/terraform-provider-gitlab/blob/375586b509db071c8aaf9f4f8b8f6d4ca6bd6359/internal/provider/helper_test.go#L102)
* [`testAccCreateUser`](https://github.com/gitlabhq/terraform-provider-gitlab/blob/375586b509db071c8aaf9f4f8b8f6d4ca6bd6359/internal/provider/helper_test.go#L127)
* [`testAccCreateGroups`](https://github.com/gitlabhq/terraform-provider-gitlab/blob/375586b509db071c8aaf9f4f8b8f6d4ca6bd6359/internal/provider/helper_test.go#L158)

You also have to create an additional one for the project environment, ...